services/leader/client.go

```diff
@@ -0,0 +1,249 @@
+
+	sc <- campaign.NewStatus(campaign.Follower)
+
+	go func() {
+		defer func() {
+			close(sc)
+			cancel()
+			c.resetCampaign()
+		}()
+
+		// Campaign blocks until elected. Once we are elected, we get a channel
+		// that's closed if our session dies.
+		ch, err := c.client.Campaign(ctx, proposeVal)
+		if err != nil {
+			sc <- campaign.NewErrorStatus(err)
+			return
+		}
+
+		sc <- campaign.NewStatus(campaign.Leader)
+		select {
+		case <-ch:
+			sc <- campaign.NewErrorStatus(election.ErrSessionExpired)
+		case <-c.resignCh:
+			sc <- campaign.NewStatus(campaign.Follower)
+		}
+	}()
+
+	return sc, nil
+}
+
+func (c *client) resign() error {
+	if c.isClosed() {
+		return errClientClosed
+	}
+
+	// if there's an active blocking call to Campaign() stop it
+	c.Lock()
+	if c.cancelFn != nil {
+		c.cancelFn()
+		c.cancelFn = nil
+	}
+	c.Unlock()
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.opts.ResignTimeout())
+	defer cancel()
+	if err := c.client.Resign(ctx); err != nil {
+		return err
+	}
+
+	// if successfully resigned and there was a campaign in Leader state cancel
+	// it
+	select {
+	case c.resignCh <- struct{}{}:
+	default:
+	}
+
+	c.resetCampaign()
+
+	return nil
+}
+
+func (c *client) leader() (string, error) {
+	if c.isClosed() {
+		return "", errClientClosed
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.opts.LeaderTimeout())
+	defer cancel()
+	ld, err := c.client.Leader(ctx)
+	if err == concurrency.ErrElectionNoLeader {
+		return ld, ErrNoLeader
+	}
+	return ld, err
+}
+
+func (c *client) startCampaign() bool {
+	c.Lock()
+	defer c.Unlock()
+
+	if c.campaigning {
+		return false
+	}
+
+	c.campaigning = true
+	return true
+}
+
+func (c *client) resetCampaign() {
+	c.Lock()
+	c.campaigning = false
+	c.Unlock()
+}
+
+// Close closes the election service client entirely. No more campaigns can be
+// started and any outstanding campaigns are closed.
+func (c *client) close() error {
+	c.Lock()
+	c.closed = true
+	c.Unlock()
+	return c.client.Close()
+}
+
+func (c *client) isClosed() bool {
+	c.RLock()
+	defer c.RUnlock()
+	return c.closed
+}
+
+// val returns the value the leader should propose based on (1) a potentially
+// empty override value and (2) the hostname of the caller (with a fallback to
+// the DefaultValue option).
+func (c *client) val(override string) string {
+	if override != "" {
+		return override
+	}
+
+	return c.opts.Hostname()
+}
+
+// elections for a service "svc" in env "test" should be stored under
+// "_ld/test/svc". A service "svc" with no environment will be stored under
+// "_ld/svc".
+func servicePrefix(sid services.ServiceID) string {
+	env := sid.Environment()
+	if env == "" {
+		return fmt.Sprintf(keyFormat, leaderKeyPrefix, sid.Name())
+	}
+
+	return fmt.Sprintf(
+		keyFormat,
+		leaderKeyPrefix,
+		fmt.Sprintf(keyFormat, env, sid.Name()))
+}
+
+func electionPrefix(sid services.ServiceID, electionID string) string {
+	eid := electionID
```

Comment:
I think it's cleaner in terms of the keys in etcd if we don't allow empty election IDs, since the keys would look like `<prefix>//<leaseID>`